### PR TITLE
Truffle option

### PR DESCRIPTION
**Pull request by FabijanC**
Along with regular files, the validation script now accepts directories and zip files (e.g. Truffle projects). Feedback is provided in the form of a structured output.

### CODE REVIEW

File: services/validation/src/ValidationService.ts [hunk 1]
```diff
@@ -46,48 +45,20 @@ export class ValidationService implements IValidationService {
...
-
-            // Case: default
-            const msg = `Invalid file(s) detected: ${util.inspect(files.files)}`;
-            this.logger.error(msg);
-            throw new Error(msg);
-        } else if (files) {
-            const truffleFiles: any[] = this.attemptTruffleExtraction(files[0]);
+    private findInputFiles(files: {data: Buffer}[]): Buffer[] {
+        if (files.length === 1) {
+            const truffleFiles = this.attemptTruffleExtraction(files[0]);
```
Comment (chriseth) [inline services/validation/src/ValidationService.ts:50]: Is there a way to make this more generic, i.e. not mention "truffle" explicitly in the function name and in general search all files for "encapsulated" metadata?

Comment (FabijanC) [reply services/validation/src/ValidationService.ts:50]: Sure, I'll make changes in that direction; just to note that a few more changes have been added to the pull request, but have not yet been published.

File: .vscode/launch.json [hunk 2]
```diff
@@ -164,6 +164,38 @@
...
+      "skipFiles": [
+        "<node_internals>/**"
+      ],
+      "program": "${workspaceFolder}/services/validation/build/index.js",
+      "preLaunchTask": "tsc: build - services/validation/tsconfig.json",
+      "outFiles": [
+        "${workspaceFolder}/services/validation/build/**/*.js"
+      ],
+      "args": [
+        "services/validation/test/truffle-example-incorrect"
```
Comment (Jakic007) [inline .vscode/launch.json:196]: This directory is missing, there is only zip of this.

File: .vscode/launch.json [hunk 3]
```diff
@@ -164,6 +164,38 @@
...
+      "skipFiles": [
+        "<node_internals>/**"
+      ],
+      "program": "${workspaceFolder}/services/validation/build/index.js",
+      "preLaunchTask": "tsc: build - services/validation/tsconfig.json",
+      "outFiles": [
+        "${workspaceFolder}/services/validation/build/**/*.js"
+      ],
+      "args": [
+        "services/validation/test/truffle-example"
```
Comment (Jakic007) [inline .vscode/launch.json:180]: This directory is actually missing .json files and should be "truffle-example-incorrect" so this returns to me that I am missing metadata.json file.

File: .vscode/launch.json (hunk 2 above)
Comment (Jakic007) [inline .vscode/launch.json:196]: Rename `truffle-example directory` to` truffle-example-incorrect` and it should be fine.

File: services/validation/src/ValidationService.ts [hunk 4]
```diff
@@ -22,72 +47,89 @@ export class ValidationService implements IValidationService {
...
+            checked.push(new CheckedContract(metadata, foundSources, missingSources, info));
         });
-        const response: CheckFileResponse = {
-            files: sources,
-            error: error
+
+        if (errorMsgMaterial.length) {
+            const msg = errorMsgMaterial.join("\n");
+            this.logger.error(msg);
+            //throw new Error(msg); if uncommented, doesn't print info about successful validations
```
Comment (Jakic007) [inline services/validation/src/ValidationService.ts:74]: Do we need this error throw?

File: .vscode/launch.json (hunk 3 above)
Comment (FabijanC) [reply .vscode/launch.json:180]: @Jakic007 I think this was caused by .gitignore specifying everything called build to be ignored. I'll fix this.

File: services/validation/src/ValidationService.ts (hunk 4 above)
Comment (FabijanC) [reply services/validation/src/ValidationService.ts]: Removed the commented line.

File: .vscode/launch.json (hunk 2 above)
Comment (FabijanC) [reply .vscode/launch.json:196]: Caused by issues in .gitignore; now resolved.

**Review by Jakic007 — APPROVED** [review]

**Review by Jakic007 — APPROVED** [review]

**Comment by Jakic007** [comment]
@chriseth if you agree, we can merge this PR to services branch and publish the new sourcify-validation package for users to test?

**Comment by chriseth** [comment]
Yes, please publish!

File: services/validation/README.md [hunk 5]
```diff
@@ -5,6 +5,6 @@
 ##### How to?
 
 * `npm install`
-* `npm run start file1.sol metadata.json`
-    * replace `file1.sol` and `metadata.json` with relative or absolute path to your sourcefiles
-* If everything went well, proceed here [Sourcify](https://verification.komputing.org/)
\ No newline at end of file
+* `npm run start path/to/file1 path/to/file2 path/to/dir path/to/zip`
+    * the script will scan through all the provided files, directories and zips searching for metadata and the belonging source files
+* If everything went well, proceed here: [Sourcify](https://verification.komputing.org/)
```
Comment (chriseth) [inline services/validation/README.md:10]: I think we could add some more test here. Perhaps an example run with output?

Comment (Jakic007) [reply services/validation/README.md:10]: Sure, we will update this file with more examples

**Pull request merged**